src/controllers/SafeNavigationBaseController.ts

```diff
@@ -0,0 +1,97 @@
+
+import { BaseController, FormSanitizeFunction } from 'app/controllers/BaseController';
+import {
+    FormSubmissionProcessor,
+    FormSubmissionProcessorConstructor
+} from 'app/controllers/processors/FormSubmissionProcessor';
+import { ApplicationData, APPEALS_KEY } from 'app/models/Appeal';
+import { getEnvOrDefault } from 'app/utils/EnvironmentUtils';
+import { PENALTY_DETAILS_PAGE_URI } from 'app/utils/Paths';
+import { Navigation } from 'app/utils/navigation/navigation';
+
+type RequestWithNavigation = Request & { navigation: Navigation }
+
+@provide(InternalProcessor)
+class InternalProcessor implements FormSubmissionProcessor {
+    constructor(@inject(SessionStore) private readonly sessionStore: SessionStore) {}
+
+    async process(request: RequestWithNavigation): Promise<void> {
+        const session = request.session.unsafeCoerce();
+        const applicationData: ApplicationData = session.getExtraData()
+            .map<ApplicationData>(data => data[APPEALS_KEY])
+            .orDefault({} as ApplicationData);
+
+        const permissions = applicationData?.navigation?.permissions || [];
+        const page = request.navigation.next(request);
+
+        if (!permissions.includes(page)) {
+            console.log('Updating page permissions');
+            session.saveExtraData(APPEALS_KEY, this.updateNavigationPermissions(applicationData, page));
+        }
+
+        await this.sessionStore
+            .store(Cookie.representationOf(session, getEnvOrDefault('COOKIE_SECRET')), session.data)
+            .run();
+    }
+
+    private updateNavigationPermissions(appealExtraData: ApplicationData, page: string): ApplicationData {
+        return {
+            ...appealExtraData,
+            navigation: {
+                permissions: [...appealExtraData?.navigation?.permissions || [], page]
+            }
+        };
+    }
+}
+
+// tslint:disable-next-line: max-classes-per-file
+export abstract class SafeNavigationBaseController<FORM> extends BaseController<FORM> {
+    protected constructor(template: string,
+                          navigation: Navigation,
+                          formSchema?: AnySchema,
+                          formSanitizeFunction?: FormSanitizeFunction<FORM>,
+                          formSubmissionProcessors?: FormSubmissionProcessorConstructor[]) {
+        super(template, navigation, formSchema, formSanitizeFunction, [
+            ...formSubmissionProcessors || [], InternalProcessor
+        ]);
+    }
+
+    async onGet(): Promise<void> {
+        const session = this.httpContext.request.session.unsafeCoerce();
+        const applicationData = session.getExtraData()
+            .map<ApplicationData>(data => data[APPEALS_KEY])
+            .orDefault({
+                navigation: {}
+            } as ApplicationData);
+
+        console.log(applicationData);
+
+        if(applicationData.navigation.permissions === undefined) {
+            console.log('Start of journey');
```

Comment:
I think it is useless log. Imagine seeing hundreds of these in logs in production instance where hundreds of users simultaneously use the service.